design/back.compatibility.md

```diff
@@ -0,0 +1,79 @@
+
+## Design Options
+
+I considered two basic approaches to implementing multi-generation support
+1. conditional checks sprinkled in whenever changes are applied("if (generation==2) then do A else do B". Pro: it's easy to understand the concept, and easy to start. Con: extremely error prone and hard to test.
+2. complete copy of the generation-specific code (i.e. almost all of the code in the dll). Pro: Minimizes risk of errors. Con: May bloat the plugin size, source tree size, and compile time. Still have to watch out for changes in the wrong version.
+
+The obvious choice here is 2. Approach 1 would causes things to break often for no reason, no matter how careful developers are.
+
+### Source Code Management
+Here are 2 options to managing the source code
+1. move all the generation-specific code into a subfolder per generation - this is where the preview generation is worked on. When finalizing a generation and starting a new preview generation, create a copy of the preview generation folder. Change the namespace in the subfolder. Add a generation-specific prefix on each source file's filename
+2. move all generation-specific code into a separate repository, and have the main repository include this repository as a submodule for each generation.
+
+* Pros for 1: Easy to use with git. 
+* Cons for 1: It's clunky to have a prefix on filenames (but makes it much harder to accidentally modify things in a final generation - many tools highlight filename but not path). It's still possible to modify the wrong generation's files in the repo.
+* Pros for 2: Makes it harder to do accidental checkins in the wrong repo - you commit into a specific version.
+* Cons for 2: git submodule workflow is clunky, not integrated well with visual git tools. While it seems like meging across generations is easier, it's not something we'll see in practice as fixing finalized generations is not something we should be doing.
+
+I recommend version 1. The pain of the git submodule workflow, and loss of productivity is not worth the minor benefits of 2
+
+### Final Binaries
+Here are two options for building the the binaries
+1. Build a single DLL that includes all generations, with a small bit of generation-shared code for connecting to a server, finding the server version, and choosing the generation-specific mre. 
+2. Build one DLLs per generation, plus a management DLL. 
+Pros for 1: it's easier to have a single output DLL than multiple, as build scripts don't need to be updated when a generation is added. Total binary size is smaller than 2, as compiler will probably optimize out many functions that are unchanged across generations. Con: Higher build time for the DLL.
+I recommend option 1. 
```

Comment:
There is a lot of power and flexibility in having a DLL per major version, as host apps are able to take only the binary versions that they want to support and have full control over the size of their dependencies that they want to include in to their final build.  Not all host apps will want to support all versions going back to V1.  I don't actually think that the assumption that the compiler will do as you think with this in Unity, as this is not a typical use case that IL2CPP would be built specifically to handle.  I think it is better to not make such an assumption in this design, and make it a factor in deciding the method by which we release or final binaries.

This is based off of an offline conversation with @eanders-MS, and I will let him chime in with his thoughts as well.

Reply:
It might be a lot simpler to put the put the versioned code into per-version DLLs, with a front-end DLL for the shared parts. One big benefit to this approach is it gives platforms the ability to choose how much backward compatibility to support without still having to take all prior runtime versions.

It would be good to know if il2cpp will do any sort of code coalescing.

Reply:
Yeah, discussing with Tom, putting a DLL per version, and a shared DLL makes sense, so I'll change this recommendation. At a later point, if total size of all DLLs is a concern, we can investigate if if merging it into a single DLL would strip a lot of unwanted code (I still think it would), but that is an optimization that we don't need to worry about right now.

We'll stick with the plan for having a folder per generation, naming the namespace differently per generation (to make it 100% unambiguous when you see a callstack). We'll create a CSProj per generation DLL, and one for the shared DLL, and then probably have a couple of different solutions - one that includes all generations (for build servers), and one that just includes shared and a specific generation's DLL (speeds up development)